Make EC code available from within the FIPS provider

mattcaswell opened the pull request:
This is built on top of #9111 so until that is merged this one will remain in WIP. Only the last 3 commits are relevant to this PR.

mattcaswell commented:
Fixup commit pushed to address a travis failure.

levitte commented:
Looks like that wasn't the only failure.

```
crypto/ec/fips-dso-ecp_nistz256.o: In function `ecp_nistz256_get_affine':
/home/travis/.ccache/tmp/ecp_nistz2.stdout.travis-20190703065617.20133.pwp246.i:(.text+0xb80): undefined reference to `ecp_nistz256_from_mont'
/home/travis/.ccache/tmp/ecp_nistz2.stdout.travis-20190703065617.20133.pwp246.i:(.text+0xbec): undefined reference to `ecp_nistz256_from_mont'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
Makefile:13275: recipe for target 'providers/fips.so' failed
make[1]: *** [providers/fips.so] Error 1
make[1]: Leaving directory '/home/travis/build/openssl/openssl'
Makefile:196: recipe for target 'tests' failed
make: *** [tests] Error 2
```

It might be because `-ansi` is used...

paulidale commented:
#9111 is merged.

mattcaswell commented:
Rebased this now that #9111 is merged. I've not looked at the travis failure yet.

mattcaswell commented:
I can't recreate it locally, but I *think* I figured out the travis failure. Fixup commit pushed.

romen commented on crypto/ec/ec_curve.c:
```diff
@@ -3090,9 +3107,10 @@ static EC_GROUP *ec_group_new_from_data(const ec_list_element curve)
+                               curve.meth != NULL ? curve.meth() : NULL);
 
-    if ((ctx = BN_CTX_new()) == NULL) {
+    if ((ctx = BN_CTX_new_ex(libctx)) == NULL) {
```
Is `ERR_R_MALLOC_FAILURE` in the line below still relevant with the new `BN_CTX_new_ex()`?

mattcaswell replied:
It's still the only reason that `BN_CTX_new_ex()` will fail, so I think it is ok.

romen commented on doc/man3/EC_GROUP_new.pod:
```diff
@@ -71,12 +75,18 @@ y^2 + xy = x^3 + ax^2 + b (where b != 0)
 EC_GROUP_new_from_ecpkparameters() will create a group from the specific PK B<params>.
 
+EC_GROUP_new is the same as EC_GROUP_new_ex() except that the library context
+used is always the default OPENSSl_CTX.
```
`OPENSSL_CTX` ?

mattcaswell replied:
I changed this to refer to the "default library context" to be consistent with how it is referred to elsewhere.

mattcaswell commented:
Fixup commit pushed addressing the documentation feedback above. My earlier fixup does seem to have resolved the travis error.

mattcaswell commented:
Now out of WIP.

mattcaswell commented:
Rebased to resolve conflicts with master.

@levitte, @romen are there any further comments on this PR?

mattcaswell commented:
Ping @levitte, @romen...any further comments?

If not @levitte - can you reconfirm?

levitte commented on include/openssl/ec.h:
```diff
@@ -422,7 +447,8 @@ size_t EC_get_builtin_curves(EC_builtin_curve *r, size_t nitems);
 const char *EC_curve_nid2nist(int nid);
 int EC_curve_nist2nid(const char *name);
-int EC_GROUP_check_named_curve(const EC_GROUP *group, int nist_only);
+int EC_GROUP_check_named_curve(const EC_GROUP *group, int nist_only,
```
You seem to add new functions with the `_ex` suffix, which is commendable considering we want to avoid API breaks,... except here. Any reason why?

romen replied:
I believe it is because `EC_GROUP_check_named_curve` is new functionality only existing in `master`

levitte replied:
Ah, ok

mattcaswell replied:
Yes - exactly.

romen commented:
> Ping @levitte, @romen...any further comments?
> 

Doing some testing, but so far everything seems fine!

romen commented:
@mattcaswell what's the plan about `crypto/ec/ecp_nistp224.c`, `crypto/ec/ecp_nistp256.c` and `crypto/ec/ecp_nistp521.c` in relation to `FIPS_MODE`?

Currently this wound not link with `enable-ec_nistp_64_gcc_128` because of references to `BN_CTX_new()`

mattcaswell commented:
> Currently this wound not link with enable-ec_nistp_64_gcc_128 because of references to BN_CTX_new()

Oh - oops. Thanks for the hint. I forgot about those files. I've fixed that how and rebased due to a conflict with master. Please take another look. The `ectest` fails with `enable-ec_nistp_64_gcc_128` but that seems to be due to @9251 rather than this PR.

romen commented:
I will have a look, we might also want to rebase this on #9474 after it's merged

mattcaswell commented:
Rebased to resolve a conflict with master.

Ping?

mattcaswell commented:
Pushed. Thanks!
